poly-synth/src/noisegate-side-chain.js

```diff
@@ -0,0 +1,212 @@
+   *                                      default has been set experimentally to
+   *                                      0.0025s.
+   * @param {Number} options.threshold Decibel level beneath which sound is
+   *                                   muted in the second channel.
+   *                                   The default is -100 dBFS.
+   */
+  constructor(context, options) {
+    if (!(context instanceof BaseAudioContext))
+      throw 'Not a valid audio context.';
+    if (!options) options = {};
+
+    this.context_ = context;
+    const bufferSize = options.bufferSize || 0;
+    this.attack = options.attack || 0;
+    this.release = options.release || 0;
+
+    // The time constant of the filter has been set experimentally to balance
+    // roughly delay for high frequency suppression.
+    let timeConstant = (typeof options.timeConstant === 'undefined') ?
+        0.0025 : options.timeConstant;
+
+    this.threshold = (typeof options.threshold === 'undefined') ?
+        -100 : options.threshold;
+
+    // Alpha controls a tradeoff between the smoothness of the
+    // envelope and its delay, with a higher value giving more smoothness at
+    // the expense of delay and vice versa.
+    this.alpha_ = this.getAlphaFromTimeConstant_(
+        timeConstant, this.context_.sampleRate);
+
+    this.noiseGateKernel_ = this.context_.createScriptProcessor(
+        bufferSize, 2, 1);
+    this.noiseGateKernel_.onaudioprocess = this.onaudioprocess_.bind(this);
+
+    // The noise gate is connected to and from by dummy input and output nodes.
+    this.input = new GainNode(this.context_);
+    this.output = new GainNode(this.context_);
+    this.input.connect(this.noiseGateKernel_).connect(this.output);
+
+    // The previous envelope level, a float representing signal amplitude.
+    this.previousLevel_ = 0;
+
+    // The last weight (between 0 and 1) assigned, where 1 means the gate
+    // is open and 0 means it is closed and the sample in the output buffer is
+    // muted. When attacking, the weight will linearly decrease from 1 to 0, and
+    // when releasing the weight linearly increase from 0 to 1.
+    this.previousWeight_ = 1.0;
+    this.channel_ = new Float32Array(this.noiseGateKernel_.bufferSize);
+    this.envelope_ = new Float32Array(this.noiseGateKernel_.bufferSize);
+    this.weights_ = new Float32Array(this.noiseGateKernel_.bufferSize);
+  }
+
+  /**
+   * Control the dynamic range of input in the second channel based on the
+   * measured level of the signal in the first channel relative to a specified
+   * threshold.
+   * @param {AudioProcessingEvent} event An Event object containing
+   *                                     2 input channels and 1 output channel.
+   */
+  onaudioprocess_(event) {
+    let inputBuffer = event.inputBuffer;
+    let envelopeChannel = inputBuffer.getChannelData(0);
+    let sideChainChannel = inputBuffer.getChannelData(1);
```

Comment:
I agree with hoch@. Not sure if that's needed at this point.  If you were doing this for the actual noise gate, this would be a requirement.